### PR TITLE
feat(store): Refund filtered and dropped attachments

### PR DESCRIPTION
**Pull request by jan-auer**
Follow-up to https://github.com/getsentry/sentry/pull/19567. Tracks outcomes for all paths in which we discard or drop attachments and refunds their quota.

Refunds are more theoretical. We currently do not have any tracked rate limits for attachments, and it's unlikely that we ever will.

#sync-getsentry

### CODE REVIEW

**Review by untitaker — APPROVED** [review]

**Review by untitaker — APPROVED** [review]

**Review by untitaker — APPROVED** [review]

**Comment by jan-auer** [comment]
This slows down `save_event` significantly enough to cause soft timeouts. Need to verify why this happens, next.

File: src/sentry/attachments/base.py [hunk 1]
```diff
@@ -41,6 +42,13 @@ def __init__(
         assert isinstance(self.type, string_types), self.type
         self.rate_limited = rate_limited
 
+        if size is not None:
+            self.size = size
+        elif data is not None:
```
Comment (untitaker) [inline src/sentry/attachments/base.py:47]: use UNINITIALIZED_DATA here

**Review by untitaker — APPROVED** [review]
we are still not sure if we fixed the perf issue but we certainly do less now

File: src/sentry/attachments/base.py (hunk 1 above)
Comment (jan-auer) [reply src/sentry/attachments/base.py]: good catch, thanks!

**Comment by jan-auer** [comment]
@untitaker this is ready for final review now. Verified correct behavior on a canary and added test coverage for the offending paths.

**Review by untitaker — APPROVED** [review]

**Pull request merged**